x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/MvPercentile.java

```diff
@@ -0,0 +1,408 @@
+        processInts(builder, position, values, percentile);
+    }
+
+    @Evaluator(extraName = "IntegerLong", warnExceptions = IllegalArgumentException.class)
+    static void process(IntBlock.Builder builder, int position, IntBlock values, long percentile) {
+        processInts(builder, position, values, percentile);
+    }
+
+    @Evaluator(extraName = "IntegerDouble", warnExceptions = IllegalArgumentException.class)
+    static void process(IntBlock.Builder builder, int position, IntBlock values, double percentile) {
+        processInts(builder, position, values, percentile);
+    }
+
+    @Evaluator(extraName = "LongInteger", warnExceptions = IllegalArgumentException.class)
+    static void process(LongBlock.Builder builder, int position, LongBlock values, int percentile) {
+        processLongs(builder, position, values, percentile);
+    }
+
+    @Evaluator(extraName = "LongLong", warnExceptions = IllegalArgumentException.class)
+    static void process(LongBlock.Builder builder, int position, LongBlock values, long percentile) {
+        processLongs(builder, position, values, percentile);
+    }
+
+    @Evaluator(extraName = "LongDouble", warnExceptions = IllegalArgumentException.class)
+    static void process(LongBlock.Builder builder, int position, LongBlock values, double percentile) {
+        processLongs(builder, position, values, percentile);
+    }
+
+    private static void processDoubles(DoubleBlock.Builder builder, int position, DoubleBlock valuesBlock, double percentile) {
+        int valueCount = valuesBlock.getValueCount(position);
+        int firstValueIndex = valuesBlock.getFirstValueIndex(position);
+
+        if (valueCount == 0) {
+            builder.appendNull();
+            return;
+        }
+
+        if (percentile < 0 || percentile > 100) {
+            throw new IllegalArgumentException("Percentile parameter must be a number between 0 and 100, found [" + percentile + "]");
+        }
+
+        if (valueCount == 1) {
+            builder.appendDouble(valuesBlock.getDouble(firstValueIndex));
+            return;
+        }
+
+        var p = percentile / 100.0;
+        var index = p * (valueCount - 1);
+        var lowerIndex = (int) index;
+        var upperIndex = lowerIndex + 1;
+        var fraction = index - lowerIndex;
+
+        if (valuesBlock.mvSortedAscending()) {
+            if (percentile == 0) {
+                builder.appendDouble(valuesBlock.getDouble(0));
+            } else if (percentile == 100) {
+                builder.appendDouble(valuesBlock.getDouble(valueCount - 1));
+            } else {
+                assert lowerIndex >= 0 && upperIndex < valueCount;
+                var result = calculateDoublePercentile(fraction, valuesBlock.getDouble(lowerIndex), valuesBlock.getDouble(upperIndex));
+                builder.appendDouble(result);
+            }
+        }
+
+        var values = new double[valueCount];
```

Comment:
Maybe move this to a @Fixed? With circuit breaking

Reply:
If you want to `@Fixed` it then it'd be a little scatch object so you can mutate the length of the array. Generally I like that.

FWIW if `percentile` is foldable and `0` you can run this as `MV_MIN`, right? Same for `100` and `MV_MAX`? That's kind of cute.

Reply:
If by "running as" you mean using the surrogate system, _I think_ it was only thought for aggregations, and only work on them right now.
I think it would be nice to have it everywhere as a generic rule for any expression, but it requires some extra work first